distsql/new_distsql.go

```diff
@@ -0,0 +1,213 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package distsql
+
+import (
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/types"
+	"github.com/pingcap/tipb/go-tipb"
+	goctx "golang.org/x/net/context"
+)
+
+var (
+	_ NewSelectResult  = &newSelectResult{}
+	_ NewPartialResult = &newPartialResult{}
+)
+
+// NewSelectResult is an iterator of coprocessor partial results.
+type NewSelectResult interface {
+	// Next gets the next partial result.
+	Next() (NewPartialResult, error)
+	// Close closes the iterator.
+	Close() error
+	// Fetch fetches partial results from client.
+	// The caller should call SetFields() before call Fetch().
+	Fetch(ctx goctx.Context)
+}
+
+// NewPartialResult is the result from a single region server.
+type NewPartialResult interface {
+	// Next returns the next rowData of the sub result.
+	// If no more row to return, rowData would be nil.
+	Next() (rowData []types.Datum, err error)
+	// Close closes the partial result.
+	Close() error
+}
+
+type newSelectResult struct {
+	label     string
+	aggregate bool
+	resp      kv.Response
+
+	results chan newResultWithErr
+	closed  chan struct{}
+
+	colLen int
```

Comment:
This name may be confused, it can be considered a column length. `rowLen` is better?